### PR TITLE
flash: add XinSheng RISC-V MCU CM32M4xxR flash program driver

### PR DESCRIPTION
**Pull request by xixijay1988**
add CMIoT XinSheng RISC-V MCU CM32M4xxR flash program driver

### CODE REVIEW

**Comment by fanghuaqi** [comment]
Did you miss flash loader source code? which might not be checked in?

File: src/flash/nor/cm32m4xxr.c [hunk 1]
```diff
@@ -0,0 +1,782 @@
...
+	}
+	return retval;
+}
+
+static int cm32m4xxr_erase(struct flash_bank *bank, unsigned first, unsigned last)
+{
+	struct target *target = bank->target;
+	// uint32_t optiondata;
+	// uint32_t obstat;
+  uint32_t ctrl_reg;
```
Comment (fanghuaqi) [inline src/flash/nor/cm32m4xxr.c]: Format your code here

File: src/flash/nor/cm32m4xxr.c [hunk 2]
```diff
@@ -0,0 +1,782 @@
...
+
+	// struct cm32m4xxr_flash_bank *cm32m4xxr_info = NULL;
+	// cm32m4xxr_info = bank->driver_priv;
+
+	if (bank->target->state != TARGET_HALTED) {
+		LOG_ERROR("Target not halted");
+		return ERROR_TARGET_NOT_HALTED;
+	}
+  //todo: write protect
```
Comment (fanghuaqi) [inline src/flash/nor/cm32m4xxr.c]: same as above

File: src/flash/nor/cm32m4xxr.c [hunk 3]
```diff
@@ -0,0 +1,782 @@
...
+		return retval;
+
+	retval = cm32m4xxr_wait_status_busy(bank, FLASH_ERASE_TIMEOUT);
+		if (retval != ERROR_OK)
+			return retval;
+
+	for (unsigned i = first; i <= last; i++) {
+
+    //clear flash pending flags
+    retval = target_write_u32(target, cm32m4xxr_get_flash_reg(bank, FMC_STS), FLASH_STS_CLRFLAG);
```
Comment (fanghuaqi) [inline src/flash/nor/cm32m4xxr.c]: same as above

File: src/flash/nor/cm32m4xxr.c [hunk 4]
```diff
@@ -0,0 +1,782 @@
...
+			return retval;
+
+		bank->sectors[i].is_erased = 1;
+
+    /* Disable the PER Bit */
+    retval = target_read_u32(target, cm32m4xxr_get_flash_reg(bank, FMC_CTRL), &ctrl_reg);
+    if (retval != ERROR_OK)
+			return retval;
+    ctrl_reg &= CR_Reset_PER;
+    retval = target_write_u32(target, cm32m4xxr_get_flash_reg(bank, FMC_CTRL), ctrl_reg);
```
Comment (fanghuaqi) [inline src/flash/nor/cm32m4xxr.c]: same as above

**Review by fanghuaqi — COMMENTED** [review]
麻烦格式化一下代码

**Comment by xixijay1988** [comment]
Sorry, because of the indent configuration difference of IDE, the code format is wrong, which has been fix now

**Pull request merged**